libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3820,6 +3919,148 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
+						(
+							*TypeProvider::withLocationIfReference(DataLocation::Storage, functionType->parameterTypesIncludingSelf().at(0)) !=
+							*TypeProvider::withLocationIfReference(DataLocation::Storage, _usingFor.typeName()->annotation().type)
+						)
+					)
+				) ||
+				(
+					functionType->parameterTypesIncludingSelf().size() == 1 &&
+					(
+						*TypeProvider::withLocationIfReference(DataLocation::Storage, functionType->parameterTypesIncludingSelf().at(0)) !=
+						*TypeProvider::withLocationIfReference(DataLocation::Storage, _usingFor.typeName()->annotation().type)
+					)
+				)
+			)
+				m_errorReporter.typeError(
+					7617_error,
+					path->location(),
+					"The function \"" + joinHumanReadable(path->path(), ".") + "\" "+
+					"needs to have one or two parameters of type " +
+					_usingFor.typeName()->annotation().type->canonicalName() +
+					" and the same data location to be used for the operator " +
+					TokenTraits::friendlyName(*operator_) +
+					"."
+				);
+			else if (
+				functionType->parameterTypesIncludingSelf().size() != 1 &&
+				functionType->parameterTypesIncludingSelf().size() != 2
+			)
+				m_errorReporter.typeError(
+					8112_error,
+					path->location(),
+					"The function \"" + joinHumanReadable(path->path(), ".") + "\" "+
+					"needs to have one or two parameters of type " +
+					_usingFor.typeName()->annotation().type->canonicalName() +
+					" and the same data location to be used for the operator " +
+					TokenTraits::friendlyName(*operator_) +
+					"."
+				);
+
+			if (
+				TokenTraits::isCompareOp(*operator_) &&
+				(functionType->returnParameterTypes().size() != 1 || *functionType->returnParameterTypes().front() != *TypeProvider::boolean())
+			)
+				m_errorReporter.typeError(
+					7995_error,
+					path->location(),
+					"The function \"" + joinHumanReadable(path->path(), ".") + "\" "+
+					"needs to return exactly one value of type bool" +
+					" to be used for the operator " +
+					TokenTraits::friendlyName(*operator_) +
+					"."
+				);
+			else if (!TokenTraits::isCompareOp(*operator_))
+			{
+				if (
+					functionType->returnParameterTypes().size() != 1 ||
+					(
+						*TypeProvider::withLocationIfReference(DataLocation::Storage, functionType->returnParameterTypes().front()) !=
+						*TypeProvider::withLocationIfReference(DataLocation::Storage, _usingFor.typeName()->annotation().type)
+					)
+				)
+					m_errorReporter.typeError(
+						7743_error,
+						path->location(),
+						"The function \"" + joinHumanReadable(path->path(), ".") + "\" "+
+						"needs to return exactly one value of type " +
+						_usingFor.typeName()->annotation().type->canonicalName() +
+						" to be used for the operator " +
+						TokenTraits::friendlyName(*operator_) +
+						"."
+					);
+				else if (*functionType->returnParameterTypes().front() != *functionType->parameterTypesIncludingSelf().front())
```

Comment:
Right. To make this work you'd have to do split off the `returnParameterCount != 1` part into a separate `if` and do both type checks only if that passes. That would be in line with the error grouping I suggested in https://github.com/ethereum/solidity/pull/12362#discussion_r976879326.

> Would that error be useful in such case? I really doubt.

That's actually a very good point. In this particular case you'd probably already get errors about params not matching `using for` and return type not matching `using for`. So an extra error about they not matching each other seems superfluous.

But in that case is this error actually useful at all? I don't think you'll ever get it without also getting a different type mismatch error. Maybe it would be better to replace it with assert that `m_errorReporter` already has some previous errors and ignore it (like we do in some places)?